### ISSUE TITLE
`ak.to_list` has stopped working on some edge cases between awkward 1.8.0 and 2.1.1

### ISSUE THREAD

**Issue by HenryDayHall** [scikit-hep/awkward#2376]
### Version of Awkward Array

2.1.1

### Description and code to reproduce

So, this is a somewhat silly edge case, but I'd rather not write exceptions for it. 

In awkward 1.8.0 I could do;
```python3
import awkward as ak
assert ak.__version__ == '1.8.0'
ak.to_list(['mystring'])
```

Sometimes I want a function to accept any list-like object, either an awkward array, a numpy array or a list, so it's convenient if `ak.to_list` works the same on all of those.

However, now in awkward 2.1.1

```python3
import awkward as ak
assert ak.__version__ == '2.1.1'
ak.to_list(['mystring'])
```
leads to
```
                                                                                                                    
    [... skipping similar frames: <listcomp> at line 80 (1487 times), _impl at line 80 (1487 times)]                
                                                                                                                    
File ~/Programs/anaconda3/envs/tree2/lib/python3.11/site-packages/awkward/operations/ak_to_list.py:80, in _impl(array)
     77     return {k: _impl(v) for k, v in array.items()}                                                          
     79 elif isinstance(array, Iterable):                 
---> 80     return [_impl(x) for x in array]              
     82 else:                                             
     83     return array                                  

File ~/Programs/anaconda3/envs/tree2/lib/python3.11/site-packages/awkward/operations/ak_to_list.py:80, in <listcomp>(.0)
     77     return {k: _impl(v) for k, v in array.items()}                                                          
     79 elif isinstance(array, Iterable):                 
---> 80     return [_impl(x) for x in array]              
     82 else:                                             
     83     return array                                  

File ~/Programs/anaconda3/envs/tree2/lib/python3.11/site-packages/awkward/operations/ak_to_list.py:49, in _impl(array)
     48 def _impl(array):                                 
---> 49     if isinstance(                                
     50         array,                                    
     51         (                                         
     52             ak.highlevel.Array,                   
     53             ak.highlevel.Record,                  
     54             ak.highlevel.ArrayBuilder,            
     55         ),                                        
     56     ):                                            
     57         return array.to_list()                    
     59     elif isinstance(array, (ak.contents.Content, ak.record.Record)):                                        

File <frozen abc>:119, in __instancecheck__(cls, instance)                                                          

RecursionError: maximum recursion depth exceeded in comparison               
```

The problem is almost certainly that a char (string length 1) is still iterable.
If it is considered reasonable to handle such a silly edge case, I will make a pull request?

**Comment by agoose77** [comment]
This is certainly a bug, and the direct fix is as you describe it! However, we have a more general fix in mind here; to eliminate the special logic in favour of a more general implementation. This issue should be closed automatically once the fix is available, thanks for filing the issue!

**Comment by agoose77** [comment]
Fixed by #2547
